src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/settings/CancelFlow.tsx

```diff
@@ -4,42 +4,182 @@
+import CancellationFlowDiscountAppliedStaticImage from "./images/CancellationFlowDiscountAppliedStaticImage.svg";
 import { useTelemetry } from "../../../../../../hooks/useTelemetry";
 import { ModalOverlay } from "../../../../../../components/client/dialog/ModalOverlay";
 import { Dialog } from "../../../../../../components/client/dialog/Dialog";
 import { Button } from "../../../../../../components/client/Button";
 import { useL10n } from "../../../../../../hooks/l10n";
 import { TelemetryButton } from "../../../../../../components/client/TelemetryButton";
+import { ExperimentData } from "../../../../../../../telemetry/generated/nimbus/experiments";
+import { onApplyCouponCode, onCheckUserHasCurrentCouponSet } from "./actions";
+import { TelemetryLink } from "../../../../../../components/client/TelemetryLink";
 
 export type Props = {
   fxaSubscriptionsUrl: string;
-  confirmationFlagEnabled: boolean;
+  enableDiscountCoupon: boolean;
+  experimentData?: ExperimentData;
+};
+
+type DiscountData = {
+  headline: string;
+  successDescription: string;
+  subtitle: string;
 };
 
 export const CancelFlow = (props: Props) => {
   const l10n = useL10n();
   const recordTelemetry = useTelemetry();
-  const currentStep = props.confirmationFlagEnabled ? "confirm" : "survey";
-  const [step, setCurrentStep] = useState<"confirm" | "survey" | "redirecting">(
-    currentStep,
-  );
+  const [step, setCurrentStep] = useState<
+    "confirm" | "survey" | "all-set" | "redirecting"
+  >("confirm");
 
   const dialogState = useOverlayTriggerState({
     onOpenChange: (isOpen) => {
       recordTelemetry("popup", isOpen ? "view" : "exit", {
         popup_id: "settings-cancel-monitor-plus-dialog",
       });
+      if (isOpen) {
+        setCurrentStep("confirm");
+      }
     },
   });
 
+  const [couponSuccess, setCouponSuccess] = useState<boolean | null>(null);
+  const [alreadyHasCouponSet, setAlreadyHasCouponSet] =
+    useState<boolean>(false);
+
+  const discountedNext3Months: DiscountData = {
+    headline: l10n.getString("settings-unsubscribe-dialog-promotion-cta", {
+      discount_percentage_num: "30%",
+      discount_duration: 3,
+    }),
+    successDescription: l10n.getString(
+      "settings-unsubscribe-dialog-promotion-description",
+      {
+        discount_percentage_num: "30%",
+        discount_duration: 3,
+      },
+    ),
+    subtitle: l10n.getString(
+      "settings-unsubscribe-dialog-promotion-limitations-apply",
+    ),
+  };
+
+  // Unless we mock out these functions, we can't test them at the moment
+  /* c8 ignore start */
+  const handleApplyCouponCode = async () => {
+    const result = await onApplyCouponCode();
+    if (result?.success) {
+      setCouponSuccess(true);
+    } else {
+      setCouponSuccess(false);
+    }
+  };
+
+  const checkCouponCode = async () => {
+    const result = await onCheckUserHasCurrentCouponSet();
+    if (typeof result.success === "boolean") {
```

Comment:
Just checking for a boolean here might not be enough. It’s safer to explicitly check if `result.success` is also truthy.